crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -248,41 +278,133 @@ where
+            (None, v2) => v2,
+            (v1, None) => v1,
+            (Some(v1), Some(v2)) => Some(v1.min(v2)),
+        };
         Ok(oldest_height)
     }
 
     #[cfg(test)]
     // TODO: This method doesn't work properly because of
     //  https://github.com/FuelLabs/fuel-core/issues/2095
     fn rollback_last_block(&self) -> StorageResult<u64> {
-        let latest_height = self
-            .db
-            .iter_all_keys::<ModificationsHistory<Description>>(Some(
-                IterDirection::Reverse,
-            ))
-            .next()
-            .ok_or(DatabaseError::ReachedEndOfHistory)??;
+        let modifications_history_migration_in_progress = self.is_migration_in_progress();
+
+        let (v2_latest_height, v1_latest_height) = self.multiversion_changes_heights(
+            IterDirection::Reverse,
+            modifications_history_migration_in_progress,
+        );
+
+        let latest_height = match (v2_latest_height, v1_latest_height) {
+            (None, None) => Err(DatabaseError::ReachedEndOfHistory)?,
+            (Some(Ok(v1)), Some(Ok(v2))) => v1.min(v2),
+            (_, Some(v1_res)) => v1_res?,
+            (Some(v2_res), _) => v2_res?,
+        };
 
         self.rollback_block_to(latest_height)?;
 
         Ok(latest_height)
     }
 
+    fn take_migration_changes(&self) -> (Changes, Option<ParkingMutexGuard<Changes>>) {
+        if self.is_migration_in_progress() {
+            let mut lock_guard = self.migration_changes.lock();
+            (std::mem::take(&mut *lock_guard), Some(lock_guard))
+        } else {
+            (Changes::default(), None)
+        }
+    }
+
+    fn add_migration_changes(
+        &self,
+        lock_guard: Option<ParkingMutexGuard<Changes>>,
+        changes: Changes,
+    ) -> StorageResult<()> {
+        match lock_guard {
+            // The only case when we did not acquire a lock_guard is because the migration is not in progress.
+            // We do not need to do anything in this case.
+            None => Ok(()),
+            Some(mut lock_guard) => {
+                let current_changes = std::mem::take(&mut *lock_guard);
+                let mut current_changes_transaction = StorageTransaction::transaction(
+                    &self.db,
+                    ConflictPolicy::Overwrite,
+                    current_changes,
+                );
+
+                StorageTransaction::transaction(
+                    &mut current_changes_transaction,
+                    ConflictPolicy::Overwrite,
+                    changes,
+                )
+                .commit()?;
+
+                *lock_guard = current_changes_transaction.into_changes();
+
+                Ok(())
+            }
+        }
+    }
+
     fn rollback_block_to(&self, height_to_rollback: u64) -> StorageResult<()> {
-        let mut storage_transaction = self.db.read_transaction();
+        let (cumulative_changes, maybe_cumulative_changes_guard) =
+            self.take_migration_changes();
```

Comment:
Could you elaborate more on why we need to change how `rollback_block_to` should work?

How migration is going shouldn't affect the work of the `rollback_block_to` method. Migration may have a state higher than the chain's height (after rollback), and it can be a problem. But we can handle this case in the `commit_changes` function and remove modifications that are higher than blockchain height.

Reply:
I follow only partially here. I was under the impression that `rollback_block_to` reverts all the changes from the current height to the one specified in the function, but this does not seem to be the case. This leaves me with the question: why is it okay to revert only the changes at the block height specified in input? Do we use this function only to rollback to the last block (which seems to be the case from the codebase)? 

Regarding your question, my flow of thought is as follows: Migration changes are piggybacked into transactions whenever they are committed to the rocksDB instance. We do this in two places: `rollback_block_to` and `<HistoricalRocksDB as TransactableStorage>::commit_changes`. In the first case we compute the last_changes by taking into account the migration changes from the in memory transaction, so even if the key for that height is being migrated from V1 to V2, the migrated changes will be captured in `last_changes` and removed. Please let me know if i am missing something. 

If the cumulative changes were not flushed here, they would be kept in memory and flush when we later commit a block. Now we could have a scenario where we rollbacked a block several time, but we did not commit the migration changes for blocks at a higher height.

Reply:
> This leaves me with the question: why is it okay to revert only the changes at the block height specified in input? Do we use this function only to rollback to the last block (which seems to be the case from the codebase)?

It is a temporary workaround because `rollback_last_block` doesn't work as expected.

> Migration changes are piggybacked into transactions whenever they are committed to the rocksDB instance.

For us it will be enough to commit changes from the migration only in `commit_changes`. 

> If the cumulative changes were not flushed here, they would be kept in memory and flush when we later commit a block. Now we could have a scenario where we rollbacked a block several time, but we did not commit the migration changes for blocks at a higher height.

I don't see any problems in keeping them in the memory until we start importing new blocks(it is okay if we need to rollback a lot of blocks. Usually it is how we use `rollback_block_to`).

Reply:
> I don't see any problems in keeping them in the memory until we start importing new blocks(it is okay if we need to rollback a lot of blocks. Usually it is how we use rollback_block_to).

My main concern in this case is a scenario that would go as follows (note that unless I have misunderstood the rollback and commit functions, this can happen): 

- Say the max height of the modification history is 100. We rollback the last block 50 times, using a strategy where we do not commit the migration changes in rocksDB. 
- During this time the migration process for all blocks completes and the changes are persisted in memory, 
- The next time the changes for a block height are committed, all the migration changes (including those at height > 50) will be committed, leaving the DB in an inconsistent state. 

You mentioned to keep track of the maximum height of the migration, but if rollbacks are not frequent a simpler alternative would be to wipe all migration changes before committing a transaction. Though this would require the lock to be kept while the commit to rocksDB happens to avoid data races.

Reply:
(Or we can simply wipe all migration changes higher than the block height to be committed, when committing)